### PR TITLE
Add devtools service

### PR DESCRIPTION
**Pull request by helin24**
This change is to start running a DevTools server for the duration of a project instead of having it associated with an application. 

Currently we don't use the daemon to serve DevTools for non-bazel projects, but this should be okay once we're using `pub global run devtools` in flutter tools.

### CODE REVIEW

File: src/io/flutter/run/daemon/DaemonApi.java [hunk 1]
```diff
@@ -404,6 +408,14 @@ RestartResult parseResult(JsonElement result) {
     }
   }
 
+  private static class DaemonShutdown extends Params<Boolean> {
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DaemonApi.java:411]: no need for an inner class. Nust write as a Lambda.

File: src/io/flutter/run/daemon/DevToolsService.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  private DevToolsInstance devToolsInstance;
+
+  @NotNull
+  public static DevToolsService getInstance(@NotNull final Project project) {
+    return ServiceManager.getService(project, DevToolsService.class);
+  }
+
+  private DevToolsService(@NotNull final Project project) {
+    this.project = project;
+    try {
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DevToolsService.java:45]: have you tested that this code gracefully handles changing your flutter project?

File: src/io/flutter/run/daemon/DevToolsService.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    if (workspace != null) {
+      final String script = workspace.getDaemonScript();
+      if (script != null) {
+        return createCommand(workspace.getRoot().getPath(), script, ImmutableList.of(), androidHome);
+      }
+
+    }
+
+    // Otherwise, use the Flutter SDK.
+    final FlutterSdk sdk = FlutterSdk.getFlutterSdk(project);
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DevToolsService.java:93]: what happens if the flutter sdk changes once this is already initialized?

File: src/io/flutter/run/daemon/DevToolsService.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        daemonApi.daemonShutdown();
+        if (!process.isProcessTerminated()) {
+          process.destroyProcess();
+        }
+      }
+    });
+  }
+
+  private static GeneralCommandLine chooseCommand(@NotNull final Project project) {
+    final String androidHome = IntelliJAndroidSdk.chooseAndroidHome(project, false);
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DevToolsService.java:80]: why are we passing android home in. That seems like it has nothing to do with serving devtools. We should either share this code with the regular code to launch the daemon or remove code that is unrelated to the launching a daemon just to serve devtools.

File: src/io/flutter/run/daemon/DevToolsService.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import static com.jetbrains.lang.dart.ide.runner.server.vmService.VmServiceWrapper.LOG;
+
+public class DevToolsService {
+  private class DevToolsServiceListener implements DaemonEvent.Listener {
+  }
+
+  @NotNull private final Project project;
+  private DaemonApi daemonApi;
+  private ProcessHandler process;
+  private DevToolsInstance devToolsInstance;
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DevToolsService.java:36]: How can a caller of this know whether the devtools instance has started yet or not?
Perhaps vend a CompletableFuture with the devToolsInstance or a boolean indicating the service has started.
Also, how is the service stopped iff a project is closed?

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 2 above)
Comment (helin24) [reply src/io/flutter/run/daemon/DevToolsService.java:48]: Hmm.. I'm not sure what this would look like, i.e. how does someone change the project? If they open a new project, the project will get its own devtools server.

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 4 above)
Comment (helin24) [reply src/io/flutter/run/daemon/DevToolsService.java]: Ah I didn't understand why this was relevant for DeviceService (where I copied this code from), but I think I can just delete this.

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 3 above)
Comment (helin24) [reply src/io/flutter/run/daemon/DevToolsService.java:101]: At this point nothing - the server will keep running and presumably there aren't many cases where users need a different version of devtools while working on a project. Are you thinking of some situations where this would be a problem?

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 1 above)
Comment (helin24) [reply src/io/flutter/run/daemon/DaemonApi.java:417]: I think I'm not doing this correctly - see above (ln 77). A class is required that implements parseResult, is there a way to do that with a lambda?

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 3 above)
Comment (jacob314) [reply src/io/flutter/run/daemon/DevToolsService.java:104]: if you changed your flutter sdk the version of the devtools server much change so you want to make sure you have the one from the new sdk not the old one.

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 2 above)
Comment (jacob314) [reply src/io/flutter/run/daemon/DevToolsService.java:48]: if you have two projects open,  having two devtools servers is fine. what I'm worried about is if you open a new project and close your existing project. do we make sure we stop the first devtools server?

Comment (helin24) [reply src/io/flutter/run/daemon/DevToolsService.java:48]: Oh right, I used `ProjectManager.getInstance().addProjectManagerListener` to listen for project closing. In there `daemon.shutdown` is called, which stops the devtools server.

File: src/io/flutter/run/daemon/DevToolsService.java (hunk 3 above)
Comment (helin24) [reply src/io/flutter/run/daemon/DevToolsService.java:104]: Hmm... I'm not sure how to manage this situation. Do we have examples elsewhere of checking whether the flutter sdk has changed?

I'm also not sure about whether users change their flutter SDK frequently enough to monitor for this? My guess was that SDK would change if users update their flutter version or if they have multiple flutter downloads, but it would also surprise me that the implementation of DevTools would change significantly enough between those versions that a user would need the newest DevTools while in the middle of working on a project. Sounds like there may be some cases I'm not thinking of?

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 1 above)
Comment (jacob314) [reply src/io/flutter/run/daemon/DaemonApi.java:417]: ah you are right. I missed the extends.

**Review by jacob314 — APPROVED** [review]
lgtm
